test/parallel_api/iterator/permutation_iterator.h

```diff
@@ -0,0 +1,263 @@
+
+    // Index of permutation iterator is based on transform iterator
+    struct transform_iterator { };
+};
+
+////////////////////////////////////////////////////////////////////////////////
+/**
+ * @param template typename TSourceIterator - source iterator type
+ * @param typename TSourceDataSize - type of source data size
+ * @param typename PermItIndexTag - tag permutation iterator base kind:
+ *          - perm_it_index_tags::counting;
+ *          - perm_it_index_tags::host;
+ *          - perm_it_index_tags::usm_shared;
+ *          - perm_it_index_tags::usm_device;
+ *          - perm_it_index_tags::transform_iterator.
+ */
+template <typename TSourceIterator, typename TSourceDataSize>
+struct test_through_permutation_iterator_data
+{
+    TSourceIterator itSource;                // First source iterator
+    const TSourceDataSize src_data_size = 0; // Source data items count
+};
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename TSourceIterator, typename TSourceDataSize, typename PermItIndexTag>
+struct test_through_permutation_iterator
+{
+};
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename TSourceIterator, typename TSourceDataSize>
+struct test_through_permutation_iterator<TSourceIterator, TSourceDataSize, perm_it_index_tags::counting>
+{
+    test_through_permutation_iterator_data<TSourceIterator, TSourceDataSize> data;
+
+    test_through_permutation_iterator(TSourceIterator itSource, const TSourceDataSize src_data_size)
+        : data{itSource, src_data_size}
+    {
+    }
+
+    template <typename TStepFunctor, typename Operand>
+    void
+    operator()(TStepFunctor stepOp, Operand op)
+    {
+        auto indexes_begin = dpl::counting_iterator<TSourceDataSize>(0);
+
+        op(dpl::make_permutation_iterator(data.itSource, indexes_begin),
+           dpl::make_permutation_iterator(data.itSource, indexes_begin) + data.src_data_size);
+    }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename TSourceIterator, typename TSourceDataSize>
+struct test_through_permutation_iterator<TSourceIterator, TSourceDataSize, perm_it_index_tags::host>
+{
+    test_through_permutation_iterator_data<TSourceIterator, TSourceDataSize> data;
+
+    test_through_permutation_iterator(TSourceIterator itSource, const TSourceDataSize src_data_size)
+        : data{itSource, src_data_size}
+    {
+    }
+
+    template <typename TStepFunctor, typename Operand>
+    void
+    operator()(TStepFunctor stepOp, Operand op)
+    {
+        ::std::vector<TSourceDataSize> indexes;
+
+        for (TSourceDataSize perm_idx_step = 1; perm_idx_step < data.src_data_size; perm_idx_step = stepOp(perm_idx_step))
+        {
+            indexes.clear();
+            for (TSourceDataSize perm_idx = 0; perm_idx < data.src_data_size; perm_idx += perm_idx_step)
+                indexes.push_back(perm_idx);
```

Comment:
To avoid deallocation/allocation/copiyng overhead on std::vector::push_back you should reserve capacity for the vector firstly.
For example, see cpprefererence 
[std::vector](http://en.cppreference.com/w/cpp/container/vector)<int, NAlloc<int>> v1;
```
        v1.reserve(max_elements); // reserves at least max_elements * sizeof(int) bytes
 
        for (int n = 0; n < max_elements; ++n)
            v1.push_back(n);
```